content/blog/mixtral-8x7b-healthcare-chatbot/index.md

```diff
@@ -0,0 +1,167 @@
+
+In this post, I’ll show you how I built a medical chatbot with Haystack 2.0, and the Mixtral 8x7B model that pulls research papers from PubMed.
+
+You can follow along [with this Colab](https://colab.research.google.com/drive/1Pl8gyfWthqoj7PTCQrteAvtOsswHvkzV). You’ll need a HuggingFace API key. [Sign up for a free account here](https://huggingface.co/join).
+
+## Challenges
+
+Building a medical chatbot presents some challenges.
+- *Lack of data*. HIPAA and other privacy regulations make it harder to find public QA datasets to fine-tune a model.
+- The human impact of providing wrong answers. 😬
+- *Staying up to date*. It takes a long time to train a LLM. By the time they are released, their knowledge is already a bit stale. Medical research breakthroughs are happening all the time. I am particularly interested in long COVID, which has both impacted humanity on a massive scale, and is the subject of ongoing research.
+
+Therefore, I decided to use a RAG pipeline to combine PubMed data with a LLM.
+
+## High Level Architecture Overview
+
+Retrieval augmented generative AI, or RAG, is a way of giving a LLM context so it can better answer questions. 
+
+You pass the LLM some documents, along with a query, and prompt the LLM to use the documents when answering the question.
+
+![](RAG.png)
+
+PubMed has up to date, trustworthy medical information so it seemed like a solid document source. Plus, there's a [PyMed wrapper for the PubMed API](https://github.com/gijswobben/pymed) that made querying easy peasy. We'll wrap this in a Haystack custom component to format the results as `Document`s so that Haystack can use them, and add some light error handling.
+
+```python
+from pymed import PubMed
+from typing import List
+from haystack import component
+from haystack import Document
+
+pubmed = PubMed(tool="Haystack2.0Prototype", email="tilde.thurium@deepset.ai")
+
+def documentize(article):
+  return Document(content=article.abstract, meta={'title': article.title, 'keywords': article.keywords})
+
+@component
+class PubMedFetcher():
+
+  @component.output_types(articles=List[Document])
+  def run(self, queries: list[str]):
+    cleaned_queries = queries[0].strip().split('\n')
+
+    articles = []
+    try:
+      for query in cleaned_queries:
+        response = pubmed.query(query, max_results = 1)
+        documents = [documentize(article) for article in response]
+        articles.extend(documents)
+    except Exception as e:
+        print(e)
+        print(f"Couldn't fetch articles for queries: {queries}" )
+    results = {'articles': articles}
+    return results
+```
+
+For the model, I went with Mixtral's 8x7b. Mixtral is a unique new kind of model that uses 8 “experts” that queries can be routed to. Each expert has 7 billion parameters, yet the queries can be pretrained quickly and have faster inference.  [This HuggingFace blog post explains MoE](https://huggingface.co/blog/moe) in more detail. 
```

Comment:
I would rephrase this. Because in reality, it's a bit more complex than 'queries' being routed. It's actually a per token routing (as per my understanding).

Instead maybe something like. Mixrtal is a unique kind of model that uses 8 "experts" and an internal "routing" mechanism that routes a token to a specific expert. This also means that during inferencing, not all parameters are used, which allows the model to response remarkably fast.